### PR TITLE
RavenDB-14634 avoid blocking record change

### PR DESCRIPTION
**Pull request by karmeli87**
- don't create TCP connection for internal nodes under lock
- don't dispose Index under lock
- don't dispose ETL under lock
- don't wait for conflicts resolution under lock

### CODE REVIEW

File: src/Raven.Server/Documents/ETL/EtlLoader.cs [hunk 1]
```diff
@@ -13,21 +13,22 @@
...
 namespace Raven.Server.Documents.ETL
 {
     public class EtlLoader : IDisposable, ITombstoneAware
     {
         private const string AlertTitle = "ETL loader";
 
-        private EtlProcess[] _processes = new EtlProcess[0];
-        private readonly HashSet<string> _uniqueConfigurationNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase); // read and modified under a lock.
+        private readonly ConcurrentSet<EtlProcess> _processes = new ConcurrentSet<EtlProcess>();
+        private readonly ConcurrentSet<string> _uniqueConfigurationNames = new ConcurrentSet<string>(StringComparer.OrdinalIgnoreCase); // read and modified under a lock.
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/EtlLoader.cs:31]: Is the comment here still accurate?

File: src/Raven.Server/Documents/ETL/EtlLoader.cs [hunk 2]
```diff
@@ -113,8 +109,10 @@ private void OnProcessRemoved(EtlProcess process)
                 if (newSqlDestinations != null && newSqlDestinations.Count > 0)
                     newProcesses.AddRange(GetRelevantProcesses<SqlEtlConfiguration, SqlConnectionString>(newSqlDestinations, ensureUniqueConfigurationNames));
 
-                processes.AddRange(newProcesses);
-                _processes = processes.ToArray();
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/EtlLoader.cs:117]: Why not keep the separate instance model?

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 3]
```diff
@@ -277,6 +277,23 @@ protected virtual void DisposeIndex()
 
                 exceptionAggregator.Execute(() => { _indexingProcessCancellationTokenSource?.Dispose(); });
 
+                if (delete)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:287]: Who calls this with `true` ?

File: src/Raven.Server/Documents/Indexes/IndexStore.cs [hunk 4]
```diff
@@ -949,7 +1000,7 @@ private void DeleteIndexInternal(Index index, bool raiseNotification = true)
 
             try
             {
-                index.Dispose();
+                index.Dispose(delete: true);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs:1021]: Why did you move the delete to here?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 5]
```diff
@@ -960,8 +970,21 @@ private void HandleInternalReplication(DatabaseRecord newRecord, List<IDisposabl
                     NodeTag = _clusterTopology.TryGetNodeTagByUrl(r).NodeTag,
                     Url = r,
                     Database = Database.Name
+                }).ToList();
+
+                Task.Run(() =>
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:991]: Cancellation token?

File: src/Sparrow/Utils/ThreadingHelper.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        public static bool InterlockedExchangeMax(ref long location, long newValue)
+        {
+            long initialValue;
+
+            do
+            {
+                initialValue = location;
+                if (initialValue >= newValue)
+                    return false;
+            } while (Interlocked.CompareExchange(ref location, newValue, initialValue) != initialValue);
```
Comment (ayende) [inline src/Sparrow/Utils/ThreadingHelper.cs:16]: Better to set the `initialvalue` to the result of `CompareExchange` here.

**Comment by ppekrol** [comment]
@karmeli87 Is is ready for merge? Can you rebase?

**Comment by karmeli87** [comment]
Closed in favor of a slimmer version: https://github.com/ravendb/ravendb/pull/12517
